### PR TITLE
feat: implement intrinsic `Int`

### PR DESCRIPTION
**Pull request by parth121101**
Towards: #492

### CODE REVIEW

**Comment by parth121101** [comment]
I have some issues for this function.
1) The file intrinics_103.f90 fails
2) Tough I have implemented this as an intrinsic elemental function I am unable to use this property.

@HarshitaKalani can you please have a look into this and if possible please let me know the problem.

File: src/libasr/pass/intrinsic_functions.h [hunk 1]
```diff
@@ -1849,6 +1850,46 @@ namespace Ibits {
...
+            return b.c2i_t(args[0], arg_type);
+        } else {
+            return nullptr;
+        }
+    }
+
+    static inline ASR::expr_t* instantiate_Int(Allocator &al, const Location &loc,
+            SymbolTable *scope, Vec<ASR::ttype_t*>& arg_types, ASR::ttype_t *return_type,
+            Vec<ASR::call_arg_t>& new_args, int64_t /*overload_id*/) {
+        declare_basic_variables("_lcompilers_anint_" + type_to_str_python(arg_types[0]));
```
Comment (HarshitaKalani) [inline src/libasr/pass/intrinsic_functions.h:1872]:
```suggestion
        declare_basic_variables("_lcompilers_int_" + type_to_str_python(arg_types[0]));
```

**Comment by HarshitaKalani** [comment]
I'll just check: @HarshitaKalani can you please have a look into this and if possible please let me know the problem.

**Comment by HarshitaKalani** [comment]
Hey @parth121101 ,
Can you please create an MRE for https://github.com/lfortran/lfortran/pull/4092#issuecomment-2132248204?

File: src/libasr/pass/intrinsic_functions.h [hunk 2]
```diff
@@ -1849,6 +1850,46 @@ namespace Ibits {
 
 } // namespace Ibits
 
+namespace Int {
+
+    static ASR::expr_t *eval_Int(Allocator &al, const Location &loc,
+            ASR::ttype_t* arg_type, Vec<ASR::expr_t*> &args, diag::Diagnostics& /*diag*/) {
+        ASRUtils::ASRBuilder b(al, loc);
+        if (is_integer(*ASRUtils::expr_type(args[0]))) {
+            return b.i_t(ASR::down_cast<ASR::IntegerConstant_t>(args[0])->m_n, arg_type);
```
Comment (HarshitaKalani) [inline src/libasr/pass/intrinsic_functions.h:1859]:
```suggestion
            return b.i2i_t(args[0], arg_type);
```
When we are passing an array as input, we cannot directly down_cast into integer. Rather use i2i_t()

File: src/libasr/pass/intrinsic_functions.h [hunk 3]
```diff
@@ -1849,6 +1850,46 @@ namespace Ibits {
...
+        } else if (is_real(*ASRUtils::expr_type(args[0]))) {
+            return b.r2i_t(args[0], arg_type);
+        } else if (is_complex(*ASRUtils::expr_type(args[0]))) {
+            return b.c2i_t(args[0], arg_type);
+        } else {
+            return nullptr;
+        }
+    }
```
Comment (HarshitaKalani) [inline src/libasr/pass/intrinsic_functions.h:1867]: I think the reason it is not behaving like elemental intrinsic is beacuse arg_type is `int32` as per `create_int` function which is an issue related to: https://github.com/lfortran/lfortran/issues/4059

File: integration_tests/CMakeLists.txt [hunk 4]
```diff
@@ -781,6 +781,7 @@ RUN(NAME intrinsics_229 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc) # derfc
 RUN(NAME intrinsics_230 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc fortran) # bessel
 RUN(NAME intrinsics_231 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc fortran) # dgamma
 RUN(NAME intrinsics_232 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc fortran) # iparity
+RUN(NAME intrinsics_233 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc fortran) # int
```
Comment (HarshitaKalani) [inline integration_tests/CMakeLists.txt:784]: We'll need to register "int" in `asr_to_fortran.cpp` file if you're using fortran label along with the test.

**Comment by Pranavchiku** [comment]
The trouble is in ASRBuilder, where we set compile time value of cast to nullptr, you'll have to handle it for every type of casting. I have done it for integer to integer, populate it for every type and I think this PR will be good then.

```diff
diff --git a/src/lfortran/semantics/ast_common_visitor.h b/src/lfortran/semantics/ast_common_visitor.h
index 276da015f..964ff0c94 100644
--- a/src/lfortran/semantics/ast_common_visitor.h
+++ b/src/lfortran/semantics/ast_common_visitor.h
@@ -5425,6 +5425,7 @@ public:
      * min([-1, 2, 3], 2, 5, [4, 4, 5], [5, -8, 7]) is broadcasted as:
      * [min(-1, 2, 5, 4, 5), min(2, 2, 5, 4, -8), min(3, 2, 5, 5, 7)]
      *
+     * int([1,2,3]) -> [int(1), int(2), int(3)], [Cast(1,1,int32,nullptr)]
     */
     void compiletime_broadcast_elemental_intrinsic(
         Vec<ASR::expr_t*> args,
@@ -5462,6 +5463,11 @@ public:
             // Call the intrinsic function for the current combination of arguments
             // result_array->m_args[i] = ASRUtils::expr_value(ASRUtils::EXPR(create_func(al, loc, intrinsic_args, diag)));
             ASR::expr_t* result = ASRUtils::expr_value(ASRUtils::EXPR(create_func(al, loc, intrinsic_args, diag)));
+            if (ASR::is_a<ASR::Cast_t>(*result)) {
+                ASR::Cast_t* cast = ASR::down_cast<ASR::Cast_t>(result);
+                if (cast->m_value == nullptr) std::cout<<"cast nullptr"<<std::endl;
+                else std::cout<<"cast not nullptr"<<std::endl;
+            }
             array_type = ASRUtils::expr_type(result);
             new_expr.push_back(al, result);
             // ASRUtils::set_ArrayConstant_value(result_array, result, i);
@@ -5471,8 +5477,20 @@ public:
                                     array_type, result_arr_type->m_dims,
                                     result_arr_type->n_dims, ASR::abiType::Source, false,
                                     result_arr_type->m_physical_type);
-        (*result_array) = ASR::down_cast<ASR::ArrayConstant_t>(ASRUtils::EXPR(ASRUtils::make_ArrayConstructor_t_util(al, (*result_array)->base.base.loc,
-                                    new_expr.p, new_expr.n, (*result_array)->m_type, (*result_array)->m_storage_format)));
+        ASR::expr_t* new_expr_ = ASRUtils::EXPR(ASRUtils::make_ArrayConstructor_t_util(al, (*result_array)->base.base.loc,
+                                    new_expr.p, new_expr.n, (*result_array)->m_type, (*result_array)->m_storage_format));
+        if (ASR::is_a<ASR::ArrayConstant_t>(*new_expr_)) {
+            (*result_array) = ASR::down_cast<ASR::ArrayConstant_t>(new_expr_);
+        } else {
+            if ( ASRUtils::expr_value(new_expr_) == nullptr ) {
+                std::cout<<"nullptr"<<std::endl;
+            } else {
+                std::cout<<"not nullptr"<<std::endl;
+            }
+            (*result_array) = ASR::down_cast<ASR::ArrayConstant_t>(ASRUtils::expr_value(new_expr_));
+        }
+        // (*result_array) = ASR::down_cast<ASR::ArrayConstant_t>(ASRUtils::EXPR(ASRUtils::make_ArrayConstructor_t_util(al, (*result_array)->base.base.loc,
+        //                             new_expr.p, new_expr.n, (*result_array)->m_type, (*result_array)->m_storage_format)));
     }
 
     std::vector<int> find_array_indices_in_args(const Vec<ASR::expr_t*>& args) {
diff --git a/src/libasr/asr_builder.h b/src/libasr/asr_builder.h
index a9873eb08..593d34739 100644
--- a/src/libasr/asr_builder.h
+++ b/src/libasr/asr_builder.h
@@ -212,7 +212,24 @@ class ASRBuilder {
     }
 
     inline ASR::expr_t* i2i_t(ASR::expr_t* x, ASR::ttype_t* t) {
-        return EXPR(ASR::make_Cast_t(al, loc, x, ASR::cast_kindType::IntegerToInteger, t, nullptr));
+        ASR::expr_t* value = nullptr;
+        if (is_value_constant(x)) {
+            LCOMPILERS_ASSERT(ASR::is_a<ASR::IntegerConstant_t>(*x));
+            int64_t val = ASR::down_cast<ASR::IntegerConstant_t>(x)->m_n;
+            int kind = extract_kind_from_ttype_t(t);
+            if (kind == 1) {
+                value = i_t(val, int8);
+            } else if (kind == 2) {
+                value = i_t(val, int16);
+            } else if (kind == 4) {
+                value = i_t(val, int32);
+            } else if (kind == 8) {
+                value = i_t(val, int64);
+            } else {
+                throw LCompilersException("Kind, " + std::to_string(kind) + " not yet supported");
+            }
+        }
+        return EXPR(ASR::make_Cast_t(al, loc, x, ASR::cast_kindType::IntegerToInteger, t, value));
     }
 
     inline ASR::expr_t* r2r_t(ASR::expr_t* x, ASR::ttype_t* t) {
```

For BOZConstant, you may start with the following diff and do consequent changes.

```diff
diff --git a/src/libasr/intrinsic_func_registry_util_gen.py b/src/libasr/intrinsic_func_registry_util_gen.py
index ac00a1d99..c1dbf3273 100644
--- a/src/libasr/intrinsic_func_registry_util_gen.py
+++ b/src/libasr/intrinsic_func_registry_util_gen.py
@@ -275,7 +275,7 @@ intrinsic_funcs_args = {
     ],
     "Int": [
         {
-            "args": [("int",), ("real",), ("complex",)],
+            "args": [("int",), ("real",), ("complex",), ("boz",)],
             "return": "int32",
             "kind_arg": True
         }
```

**Comment by certik** [comment]
I think that we should not add "boz" as a new type. It's not a type in ASR and these ASR instrinsics should just accept ASR types.

What exactly is the problem with BOZ constants that they don't work? They should be expressions of type integer or reals, so I would think they should just work.

**Comment by parth121101** [comment]
I am trying to add support for BOZ constant for this function. For this I was required to create a function that would be checking the type of this BoZ constant. I am encountering the following warning.
```
/Users/parth1211/repos/lfortran/src/libasr/../libasr/asr.h:32:26: warning: comparison of different enumeration types ('exprType' and 'ttypeType') [-Wenum-compare]
    return T::class_type == x.type;
           ~~~~~~~~~~~~~ ^  ~~~~~~
/Users/parth1211/repos/lfortran/src/libasr/../libasr/asr_utils.h:2087:17: note: in instantiation of function template specialization 'LCompilers::ASR::is_a<LCompilers::ASR::IntegerBOZ_t, LCompilers::ASR::ttype_t>' requested here
    return ASR::is_a<ASR::IntegerBOZ_t>(*type_get_past_array(type_get_past_allocatable(type_get_past_pointer(&x))));
```
It says two different enumeration types are being compared. I am unable to figure out how there are two different types being compared.

**Comment by certik** [comment]
The last error comes from using `ASR::is_a` incorrectly, `IntegerBOZ` is an `expr`, so `type_get_past_array` must be an `expr`, but I think it's a `ttype`, so you can't be calling `is_a` on it for an expr type.

**Comment by certik** [comment]
I think you are treating `BOZ` as a type, but it's not a type. It's an expression, of a type, say, integer. So the current setup should work. But since it doesn't, there is a bug somewhere. We need to figure out why it doesn't work.

**Comment by Pranavchiku** [comment]
@parth121101 I would ask you to please post an MRE.

**Comment by Pranavchiku** [comment]
The trouble is that we right now set type to nullptr for `IntegerBoZ`:

```clojure
(Print
    [(IntegerBOZ
        65
        Hex
        ()
    )]
    ()
    ()
)
```

one way is to set it to `(Integer 4)` but this would lead to lot of conflicts while downcasting to IntegerConstant vs IntegerBOZ

**Comment by certik** [comment]
Couple things to fix:
* Always set a type of all expr expressions. 
* Is IntegerBOZ always of type integer?
* Should IntegerBOZ be folded into IntegerConstant? That's what it is, right? Maybe we can add an argument to IntegerConstant to store the information how the constant was presented in the surface language (hex, octal, decimal, binary).

**Comment by parth121101** [comment]
Continuing at: #4186

**Comment by HarshitaKalani** [comment]
Closing as implemented in https://github.com/lfortran/lfortran/pull/5029

**Pull request closed without merging**